content/src/logic/third-party-contract-registry.ts

```diff
@@ -0,0 +1,128 @@
+  ERC721 = 'erc721',
+  ERC1155 = 'erc1155',
+  UNKNOWN = 'unknown'
+}
+
+export function loadCacheFile(file: string): Record<string, ContractType> {
+  try {
+    if (!fs.existsSync(file)) {
+      saveCacheFile(file, {})
+    }
+    const fileContent = fs.readFileSync(file, 'utf-8')
+    return JSON.parse(fileContent)
+  } catch (_) {
+    return {}
+  }
+}
+
+export function saveCacheFile(file: string, data: any): void {
+  const jsonData = JSON.stringify(data, null, 2)
+  fs.writeFileSync(file, jsonData, 'utf-8')
+}
+
+export type ThirdPartyContractRegistry = {
+  isErc721(contractAddress: ContractAddress): boolean
+  isErc1155(contractAddress: ContractAddress): boolean
+  isUnknown(contractAddress: ContractAddress): boolean
+  ensureContractsKnown(contractAddresses: ContractAddress[]): Promise<void>
+}
+
+export async function createThirdPartyContractRegistry(
+  logs: ILoggerComponent,
+  provider: HTTPProvider,
+  network: 'mainnet' | 'sepolia' | 'polygon' | 'amoy',
+  storageRoot: string
+): Promise<ThirdPartyContractRegistry> {
+  const logger = logs.getLogger('contract-registry')
+
+  const requestManager = new RequestManager(provider)
+  const erc721ContractFactory = new ContractFactory(requestManager, erc721Abi)
+  const erc1155ContractFactory = new ContractFactory(requestManager, erc1155Abi)
+
+  const file = path.join(storageRoot, `third-party-contracts-${network}.json`)
+  const data: Record<ContractAddress, ContractType> = loadCacheFile(file)
+
+  function isErc721(contractAddress: ContractAddress): boolean {
+    return data[contractAddress.toLowerCase()] === ContractType.ERC721
+  }
+
+  function isErc1155(contractAddress: ContractAddress): boolean {
+    return data[contractAddress.toLowerCase()] === ContractType.ERC1155
+  }
+
+  function isUnknown(contractAddress: ContractAddress): boolean {
+    return data[contractAddress.toLowerCase()] === ContractType.UNKNOWN
+  }
+
+  async function checkIfErc721(contractAddress: ContractAddress): Promise<boolean> {
+    // ERC-721 checks
+    const contract: any = await erc721ContractFactory.at(contractAddress)
+    try {
+      const r = await sendSingle(provider, await contract.ownerOf.toRPCMessage(0))
+
+      if (r.error?.code === 3) {
+        // NFT id doesn't exist, but it is an ERC-721
+        return true
+      }
+      if (!r.result) {
+        return false
+      }
+      return !!contract.ownerOf.unpackOutput(toData(r.result))
+    } catch (_) {
+      return false
+    }
+  }
+
+  async function checkIfErc1155(contractAddress: ContractAddress): Promise<boolean> {
+    // ERC-1155 checks
+    const contract: any = await erc1155ContractFactory.at(contractAddress)
+
+    try {
+      const r = await sendSingle(provider, await contract.balanceOf.toRPCMessage(contract.address, 0))
+
+      if (!r.result) {
+        return false
+      }
+      return !!contract.balanceOf.unpackOutput(toData(r.result))
+    } catch (_) {
+      return false
+    }
+  }
```

Comment:
WDYT about creating a new function `checkContractType` (maybe) with the shared logic? (non-blocker IMO)

Reply:
I'm not really sure there's much to be reused... other than the try-catch, the rest has subtle differences in each line. Open to a suggestion, may be I'm not seeing it.

Reply:
Yes, I think it's not worth the new abstraction. I thought they shared much more logic, but it's mostly code.

This was the only thing that I could think of:

```ts
 async function checkContractType(
  contractFactory: ContractFactory,
  contractAddress: ContractAddress,
  responseValidator: (response: any, contract: any, method: string) => boolean,
  method: string,
  args: any[]
): Promise<boolean> {
  try {
    const contract = await contractFactory.at(contractAddress);
    const response = await sendSingle(
      provider,
      await contract[method].toRPCMessage(...args)
    );

    return responseValidator(response, contract, method);
  } catch (error) {
    console.error(`Error checking contract type for ${contractAddress}`, error);
    return false;
  }
}

function defaultValidation(
  response: any,
  contract: any,
  method: string
): boolean {
  if (!response.result) return false;
  return !!contract[method].unpackOutput(toData(response.result));
}

async function checkIfErc721(
  contractFactory: ContractFactory,
  contractAddress: ContractAddress
): Promise<boolean> {
  const responseValidator: (
    response: any,
    contract: any,
    method: string
  ) => boolean = (response, contract) => {
    // Custom check for ERC721 (if any)
    if (response.error?.code === 3) {
      return true; // NFT id doesn't exist, but it's still an ERC721
    }
    return defaultValidation(response, contract, "ownerOf");
  };

  return checkContractType(
    contractFactory,
    contractAddress,
    responseValidator,
    "ownerOf",
    [0]
  );
}

async function checkIfErc1155(
  contractFactory: ContractFactory,
  contractAddress: ContractAddress
): Promise<boolean> {
  return checkContractType(
    contractFactory,
    contractAddress,
    defaultValidation,
    "balanceOf",
    [contractAddress, 0]
  );
}

```

Definitely not worth the added complexity 🙏